class/track/RaceTrack.js

```diff
@@ -337,59 +340,15 @@ export class RaceTrack extends Track {
-        border.y = Math.floor(border.y / this.gridSize);
-        let DI = [],
-            x, y, key;
-        for (let x = center.x; x <= border.x; x++) {
-            for (let y = center.y; y <= border.y; y++) {
-                if (this.grid[x] !== undefined && this.grid[x][y] !== undefined) {
-                    if (this.grid[x][y].lines.length > 0 || this.grid[x][y].scenery.length > 0) {
-                        DI[key = x + '_' + y] = 1;
-                        if (this.cache[key] === undefined) {
-                            let el = this.cache[key] = document.createElement('canvas');
-                            el.width = this.gridSize * this.zoomFactor;
-                            el.height = this.gridSize * this.zoomFactor;
-                            let graphic = el.getContext('2d');
-                            graphic.lineCap = 'round';
-                            graphic.lineWidth = Math.max(2 * this.zoomFactor, 0.5);
-                            graphic.strokeStyle = '#aaa';
-                            for (let i = 0, l = this.grid[x][y].scenery.length; i < l; i++) {
-                                this.grid[x][y].scenery[i].render(graphic, x * this.gridSize * this.zoomFactor, y * this.gridSize * this.zoomFactor);
-                            }
-                            graphic.strokeStyle = '#000';
-                            if (shadeLines) {
-                                graphic.shadowOffsetX =
-                                    graphic.shadowOffsetY = 2;
-                                graphic.shadowBlur = Math.max(2, 10 * this.zoomFactor);
-                                graphic.shadowColor = '#000';
-                            }
-                            for (let i = 0, l = this.grid[x][y].lines.length; i < l; i++) {
-                                this.grid[x][y].lines[i].render(graphic, x * this.gridSize * this.zoomFactor, y * this.gridSize * this.zoomFactor);
-                            }
-                        }
-                        drawer.drawImage(this.cache[key], Math.floor(canvas.width / 2 - this.camera.x * this.zoomFactor + x * this.gridSize * this.zoomFactor), Math.floor(canvas.height / 2 - this.camera.y * this.zoomFactor + y * this.gridSize * this.zoomFactor));
-                    }
-                    drawer.setProperty('strokeStyle', '#000');
-                    for (let i = 0, l = this.grid[x][y].objects.length; i < l; i++) {
-                        this.grid[x][y].objects[i].render();
-                    }
-                }
-            }
-        }
-        for (let Ay in this.cache) {
-            if (DI[Ay] === undefined) {
-                delete this.cache[Ay];
-            }
-        }
+        
+        this.drawGridBoxes(drawer);
+
+        // Don't show text or tools if you're making a thumbnail.
         if (canvas.width === 250) {
             return;
         }
-
+        
+        // Draw tools (crosshairs, eraser, powerups)
```

Comment:
I think it would be a good idea to eventually move just about everything in that function. (put each tool in their own class, restructure the render function to be mostly method calls, possibly put the main draw function in a separate class from Track, since more than just the track is being drawn.) This is a big task, and i'm not sure if you would agree, so I only did a few steps

Reply:
I was planning to make a TrackRenderer mixin so it seems right